src/main/java/functions/CheckFunction.kt

```diff
@@ -0,0 +1,22 @@
+package functions
+
+import tree.TargetTreeNode
+
+fun checkFunction(expression: VisitorFunction): VisitorFunction = { targetTreeNode ->
+    var currentNode = targetTreeNode
+    var listAnswer = listOf<List<TargetTreeNode>>()
+    val allAnswers = mutableListOf<VisitorFunctionResponse>()
+    while (currentNode != null) {
+        if (!listAnswer.any { it.any { targetNode -> targetNode == currentNode } }) {
+            val invoke = expression.invoke(currentNode)
+            allAnswers.add(invoke)
+            listAnswer = invoke.matches
+        }
+        currentNode = currentNode.moveToRightSibling()
+    }
+    if (allAnswers.any { it.hasMatch }) {
+        VisitorFunctionResponse(allAnswers.flatMap { it.matches }, true)
```

Comment:
There is a bug in this lines, specifically the last `if`. It returns every answer, regardless of `hasMatch` as long as one of the answers is correct. I propose the next code, that changes the approach but fixes the bug and should save some operations.

```suggestion
    val matchedAnswers = mutableListOf<VisitorFunctionResponse>()
    while (currentNode != null) {
        if (!listAnswer.any { it.any { targetNode -> targetNode == currentNode } }) {
            val invoke = expression.invoke(currentNode)
            if (invoke.hasMatch) {
                matchedAnswers.add(invoke)
                listAnswer = invoke.matches
            }
        }
        currentNode = currentNode.moveToRightSibling()
    }
    if (matchedAnswers.isNotEmpty()) {
        VisitorFunctionResponse(matchedAnswers.flatMap { it.matches }, true)
```